aws/protocol/query/encoder_test.go

```diff
@@ -0,0 +1,50 @@
+package query
+
+import (
+	smithytesting "github.com/awslabs/smithy-go/testing"
+	"testing"
+)
+
+func TestEncodeObject(t *testing.T) {
+	encoder := NewEncoder()
+	encoder.Object().Key("foo").String("bar")
+	smithytesting.AssertURLFormEqual(t, []byte(`foo=bar`), encoder.Bytes())
+}
+
+func TestEncodeNestedObject(t *testing.T) {
+	encoder := NewEncoder()
+	encoder.Object().Key("foo").Object().Key("bar").String("baz")
+	smithytesting.AssertURLFormEqual(t, []byte(`foo.bar=baz`), encoder.Bytes())
+}
+
+func TestEncodeList(t *testing.T) {
+	encoder := NewEncoder()
+	list := encoder.Object().Key("list").Array("spam")
+	list.Value().String("spam")
+	list.Value().String("eggs")
+	smithytesting.AssertURLFormEqual(t, []byte(`list.spam.1=spam&list.spam.2=eggs`), encoder.Bytes())
+}
+
+func TestEncodeFlatList(t *testing.T) {
+	encoder := NewEncoder()
+	list := encoder.Object().FlatKey("list").Array("spam")
+	list.Value().String("spam")
+	list.Value().String("eggs")
+	smithytesting.AssertURLFormEqual(t, []byte(`list.1=spam&list.2=eggs`), encoder.Bytes())
+}
+
+func TestEncodeMap(t *testing.T) {
+	encoder := NewEncoder()
+	mapValue := encoder.Object().Key("map").Map("key", "value")
+	mapValue.Key("bar").String("baz")
+	mapValue.Key("foo").String("bin")
+	smithytesting.AssertURLFormEqual(t, []byte(`map.entry.1.key=bar&map.entry.1.value=baz&map.entry.2.key=foo&map.entry.2.value=bin`), encoder.Bytes())
+}
+
+func TestEncodeFlatMap(t *testing.T) {
+	encoder := NewEncoder()
+	mapValue := encoder.Object().FlatKey("map").Map("key", "value")
+	mapValue.Key("bar").String("baz")
+	mapValue.Key("foo").String("bin")
+	smithytesting.AssertURLFormEqual(t, []byte(`map.1.key=bar&map.1.value=baz&map.2.key=foo&map.2.value=bin`), encoder.Bytes())
+}
```

Comment:
Nit: These tests can be a single test function that has different cases in form of a  `map[string]struct{...}`. This is generally the pattern followed by the SDK. 

An example would be : https://github.com/awslabs/smithy-go/blob/012806b992a232f8c0805c5a6600bc4fb8142985/testing/document_test.go#L8

Reply:
I wasn't sure how to square the fact that I'm building everything up and then having to call `.Bytes()` at the end, so I can't just make it one line.

Reply:
The building part can have a member `func(encoder Encoder) []byte`. Not an important change though, just consistent.

Reply:
If you were to use [table testing]() with these cases to squash a couple different tests together into a group it could look something like:

```go
func TestEncode(t *testing.T) {
     cases := map[string]struct{
          Encode func(*Encoder) error
          Expect []byte
     } {
         "list": {
               Encode: func(e Encoder) error {
                    list := e.Object().Key("list").Array("spam")
                    list.Value().String("spam")
                    list.Value().String("eggs")
                    return e.Encode()
               },
               Expect: []byte(`list.spam.1=spam&list.spam.2=eggs`),
         },
         "flat list": {
               Encode: func(e Encoder) error {
                   list := encoder.Object().FlatKey("list").Array("spam")
                   list.Value().String("spam")
                   list.Value().String("eggs")
                    return e.Encode()
               },
               Expect: []byte(`list.1=spam&list.2=eggs`),
         },
        // ...
     }

     for name, c := range cases {
         t.Run(name, func(t *testing.T) {
               var buff bytes.Buffer
	       encoder := NewEncoder(&buff)

               if err := c.Encode(encoder); err != nil {
                        t.Fatalf("failed to encode, %v", err)
               }

                smithytesting.AssertURLFormEqual(t, c.Expect, buff.Bytes())
          })
     }
}